src/convnets/resnets/core.jl

```diff
@@ -0,0 +1,331 @@
+    
+      + `:default`: Builds a stem based on the default ResNet stem, which consists of a single
+        7x7 convolution with stride 2 and a normalisation layer followed by a 3x3 max pooling
+        layer with stride 2.
+      + `:deep`: This borrows ideas from other papers (InceptionResNet-v2, for example) in using
+        a deeper stem with 3 successive 3x3 convolutions having normalisation layers after each
+        one. This is followed by a 3x3 max pooling layer with stride 2.
+      + `:deep_tiered`: A variant of the `:deep` stem that has a larger width in the second
+        convolution. This is an experimental variant from the `timm` library in Python that
+        shows peformance improvements over the `:deep` stem in some cases.
+
+  - `inchannels`: The number of channels in the input.
+  - `replace_pool`: Set to true to replace the max pooling layers with a 3x3 convolution + normalization with a stride of two.
+  - `norm_layer`: The normalisation layer used in the stem.
+  - `activation`: The activation function used in the stem.
+"""
+function resnet_stem(stem_type::Symbol = :default; inchannels::Integer = 3,
+                     replace_pool::Bool = false, activation = relu,
+                     norm_layer = BatchNorm, revnorm::Bool = false)
+    @assert stem_type in [:default, :deep, :deep_tiered]
+    "Stem type must be one of [:default, :deep, :deep_tiered]"
+    # Main stem
+    deep_stem = stem_type == :deep || stem_type == :deep_tiered
+    inplanes = deep_stem ? stem_width * 2 : 64
+    # Deep stem that uses three successive 3x3 convolutions instead of a single 7x7 convolution
+    if deep_stem
+        if stem_type == :deep
+            stem_channels = (stem_width, stem_width)
+        elseif stem_type == :deep_tiered
+            stem_channels = (3 * (stem_width ÷ 4), stem_width)
+        end
+        conv1 = Chain(conv_norm((3, 3), inchannels => stem_channels[1], activation;
+                                norm_layer, revnorm, stride = 2, pad = 1, bias = false)...,
+                      conv_norm((3, 3), stem_channels[1] => stem_channels[2], activation;
+                                norm_layer, pad = 1, bias = false)...,
+                      Conv((3, 3), stem_channels[2] => inplanes; pad = 1, bias = false))
+    else
+        conv1 = Conv((7, 7), inchannels => inplanes; stride = 2, pad = 3, bias = false)
+    end
+    bn1 = norm_layer(inplanes, activation)
+    # Stem pooling
+    stempool = replace_pool ?
+               Chain(conv_norm((3, 3), inplanes => inplanes, activation; norm_layer,
+                               revnorm,
+                               stride = 2, pad = 1, bias = false)...) :
+               MaxPool((3, 3); stride = 2, pad = 1)
+    return Chain(conv1, bn1, stempool)
+end
+
+resnet_planes(stage_idx::Integer) = 64 * 2^(stage_idx - 1)
+
+function basicblock_builder(block_repeats::Vector{<:Integer}; inplanes::Integer = 64,
```

Comment:
This probably deserves a docstring?

Reply:
Yep, docstrings are pending. I need to give them some time but I was catching up on the other models - I'll open a follow up PR if that's okay